circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/eccvm/msm_builder.hpp

```diff
@@ -0,0 +1,263 @@
+    {
+        // when we define our point lookup table, we have 2 write columns and 4 read columns
+        // when we perform a read on a given row, we need to increment the read count on the respective write column by
+        // 1 we can define the following struture: 1st write column = positive 2nd write column = negative the row
+        // number is a function of pc and slice value row = pc_delta * rows_per_point_table + some function of the slice
+        // value pc_delta = total_number_of_muls - pc std::vector<std::array<size_t, > point_table_read_counts;
+        const size_t table_rows = static_cast<size_t>(total_number_of_muls) * 8;
+        point_table_read_counts[0].reserve(table_rows);
+        point_table_read_counts[1].reserve(table_rows);
+        for (size_t i = 0; i < table_rows; ++i) {
+            point_table_read_counts[0].emplace_back(0);
+            point_table_read_counts[1].emplace_back(0);
+        }
+        const auto update_read_counts = [&](const size_t pc, const int slice) {
+            // When we compute our wnaf/point tables, we start with the point with the largest pc value.
+            // i.e. if we are reading a slice for point with a point counter value `pc`,
+            // its position in the wnaf/point table (relative to other points) will be `total_number_of_muls - pc`
+            const size_t pc_delta = total_number_of_muls - pc;
+            const size_t pc_offset = pc_delta * 8;
+            bool slice_negative = slice < 0;
+            const int slice_row = (slice + 15) / 2;
+
+            const size_t column_index = slice_negative ? 1 : 0;
+
+            if (slice_negative) {
+                point_table_read_counts[column_index][pc_offset + static_cast<size_t>(slice_row)]++;
+            } else {
+                // 8 maps to 7
+                // 15 maps to 0
+
+                // 15 - x
+                point_table_read_counts[column_index][pc_offset + 15 - static_cast<size_t>(slice_row)]++;
+            }
+            // slice : row
+            // -15 : 0
+            // -13 : 1
+            // -11 : 2
+            // -9 : 3
+            // -7 : 4
+            // -5 : 5
+            // -3 : 6
+            // -1 : 7
+            // 1 : 8
+            // 3 : 9
+            // 5 : 10
+            // 7 : 11
+            // 9 : 12
+            // 11 : 13
+            // 13 : 14
+            // 15 : 15
+        };
+        std::vector<MSMState> msm_state;
+        // start with empty row (shiftable polynomials must have 0 as first coefficient)
+        msm_state.emplace_back(MSMState{});
+        uint32_t pc = total_number_of_muls;
+        AffineElement accumulator = CycleGroup::affine_point_at_infinity;
+
+        for (const auto& msm : msms) {
+            const size_t msm_size = msm.size();
+
+            const size_t rows_per_round = (msm_size / ADDITIONS_PER_ROW) + (msm_size % ADDITIONS_PER_ROW != 0 ? 1 : 0);
```

Comment:
What are these rows and rounds?

Reply:
I will delete these and add a more explanatory comment